Preps for working on a Python-3 version

janwijbrand opened the pull request:
The preparations comprise:

* Cleanups
* Bugfix in `ImportDatabase()`
* Starting on test coverage.

janwijbrand committed:
gitignore, fix typo

janwijbrand committed:
cleanups

janwijbrand committed:
cleanups

janwijbrand committed:
add testing

janwijbrand committed:
from foo import bar as baz

janwijbrand committed:
test for attribute assignment code path

janwijbrand committed:
Add a comment for this test

janwijbrand committed:
On directory

janwijbrand committed:
no python 3 yet

janwijbrand committed:
update changlog

janwijbrand committed:
update changlog

janwijbrand committed:
Make tests for cases that are reported in the README to trigger false positives. We will drop Python 2 support in the next major version.

janwijbrand commented:
If this PR is acceptable, I can merge it and release a last `importchecker` version that works (only) on Python 2. Work for a Python 3 version is underway, based on this work.

janwijbrand committed:
Add environment classifier

janwijbrand committed:
An acicdent with a keyboard

janwijbrand committed:
imports in function bodies

mgedmin commented on CHANGES.txt:
```diff
@@ -4,8 +4,15 @@ Changes
 ----------------
 
-- Nothing changed yet.
+* NOTE: this is the last version to officially support Python 2.
```
Perhaps "to officially support Python 2 only", or do you really intend to go Python 3-only in the future?

janwijbrand replied:
see comment below.

mgedmin commented on CHANGES.txt:
```diff
@@ -4,8 +4,15 @@ Changes
 
+* Start in test coverage.
+
+* Fix a bug in `ImportDatabase()`.
```
That is not very informative.

janwijbrand replied:
fixed

mgedmin commented on CHANGES.txt:
```diff
@@ -4,8 +4,15 @@ Changes
+
+* Fix a bug in `ImportDatabase()`.
+
+* Cleanups.
```
(I would not bother mentioning externally-invisible details like this in a changelog.)

janwijbrand replied:
fixed

mgedmin commented on README.txt:
```diff
@@ -49,14 +49,11 @@ the developed codebase::
-* No Python-3 support yet.
-
-* Case like::
+* Python-2 support will be dropped in the next major version.
```
Oh, you do intend to drop Python 2 support!  I did not expect that.

janwijbrand replied:
The reason being: the Python 3 version will use the `ast` module. Even though Python 2 has the `ast` module too, the tree structure is different between Python 2 and Python 3. In informal testing I can see those differences do not affect what `importchecker` is interested in, but I'm not sure I can guarantee that nor  am I sure how to make the codebase such that it can deal with both tree structures. There's also a time constraint on my side :/

mgedmin replied:
I'm the maintainer of https://github.com/mgedmin/findimports which uses the `ast` module to, essentially, do what you're doing here in importchecker.

I support both Python 2 and Python 3 from a single codebase.  I do not remember that presenting any difficulties.

The difficult part was switching from the `compiler` module to the `ast` module (on both Python 2 and Python 3): https://github.com/mgedmin/findimports/commit/96f0f2386b01b5265ab36142aa05d4f3f4bd956f

janwijbrand replied:
@mgedmin thanks for the review and comments. `findimports` seems interesting. I wonder if `importchecker`could be replaced by it. Still, your "switch" to `ast` looks very familiar to what I did in my Python 3 branch. I was a bit put off claiming Python 2 *and* 3 support in the same code base when reading up on the differences in the syntax trees between the two (for example here http://joao.npimentel.net/2015/07/23/python-2-vs-python-3-ast-differences).

That said, I could tone down the "I'm dropping Python 2"-statement and say "This seems to still work in Python 2." (as informal testing thus far indeed seems to indicate).

mgedmin replied:
`findimports` has known bugs I haven't had time to fix for years (e.g. it doesn't understand code like `import foo.bar; foo.bar.baz()`).  Today I use `flake8` to discover unused imports.

The "Seems to still work in Python 2" bit could be easily verified with [tox](https://pypi.python.org/pypi/tox) or Travis CI, if you trust your test suite ;)

janwijbrand replied:
I don't trust my test suite TBH. I added it after the fact for code someone else wrote years and years ago :) Still, I added the test suite specifically to have some sort of baseline for the Python 3 port. I'll add a tox and travis config. We use `flake8` too, however `importchecker` catches certain import issues that `flake8` does not.

mgedmin commented on src/importchecker/importchecker.py:
```diff
@@ -18,7 +18,7 @@
 
 $Id$
```
My fingers are itching to mass-remove these Subversion crumbs.

janwijbrand replied:
fixed

mgedmin commented on src/importchecker/importchecker.py:
```diff
@@ -290,14 +290,14 @@ def getModulesImportingNameFrom(self, name, module):
     except IndexError:
-        print "No path supplied"
+        print("No path supplied", file=stdout)
         sys.exit(1)
```
FWIW you can `sys.exit("No path supplied")` and it will print the message to stderr and exit with status 1.

janwijbrand replied:
This way I can use an alternative output stream for testing. But I didn't add a test yet. Will do that now.

mgedmin reviewed (approved):
LGTM

janwijbrand committed:
Process comments

https://github.com/zopefoundation/importchecker/pull/2#discussion_r169311130
https://github.com/zopefoundation/importchecker/pull/2#discussion_r169311217

janwijbrand committed:
remove Subversion crumbs

janwijbrand committed:
the name of the utility

janwijbrand committed:
test 'no path supplied'

mgedmin commented:
BTW it might be a good idea to set up Travis CI.  If you discover you need more than a `.travis.yml` file in the repo, I have sufficient GitHub organization permissions to enable Travis CI hooks.

mgedmin commented on src/importchecker/tests/test_importchecker.py:
```diff
@@ -7,6 +7,19 @@
+        self.assertEqual(
+            'No path supplied\n',
+            output.getvalue())
+        sys.argv[:] = orig
```
If the assertion fails, the cleanup code will not run and sys.argv will remain changed.

(Personally I like `with mock.patch('sys.argv', []): ...`, but if you don't want to add mock to the dependencies, you could write a `self.call_main()` helper that uses try/finally or something like that.)

janwijbrand replied:
`mock` is ok I guess. Fixed it.

janwijbrand committed:
`mock` as testing requirement

janwijbrand committed:
now that we depend on `mock` we'll have and `.eggs` subdir when running `python2 ./setup.py test`

janwijbrand committed:
mock sys.argv.

janwijbrand committed:
get tests to work with zope.testrunner and tox and travis

janwijbrand commented:
@mgedmin I think we need configure this repo to have Travis pick it up.

mgedmin commented:
Yeah, https://travis-ci.org/zopefoundation/importchecker did a
![ekrano nuotrauka is 2018-02-21 15-11-05](https://user-images.githubusercontent.com/159967/36481773-781d9844-1719-11e8-8299-67d6a9be661d.png)

so I hit the Activate button and thing should start happening on next push!

janwijbrand committed:
update changelog

mgedmin commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+    - pip install -U pip setuptools
+    - pip install -U -e .[test]
+script:
+    - python -m zope.testrunner --test-path=src {posargs:-vc}
```
No reason to use `python -m zope.testrunner` instead of `zope-testrunner`.

You may want to consider `coverage run -m zope.testrunner`, and then we can enable coveralls.io integration as well.

mgedmin commented on buildout.cfg:
```diff
@@ -1,9 +1,16 @@
 [importchecker]
 recipe = zc.recipe.egg
-eggs = importchecker
+eggs = importchecker[test]
```
Why `[test]` for the non-test part?

mgedmin commented on tox.ini:
```diff
@@ -0,0 +1,9 @@
+    py27
+[testenv]
+commands =
+    python -m zope.testrunner --test-path=src {posargs:-vc}
```
I would also use `zope-testrunner --test-path=src {postargs:-vc}` here.

mgedmin commented on tox.ini:
```diff
@@ -0,0 +1,9 @@
+    python -m zope.testrunner --test-path=src {posargs:-vc}
+deps =
+    .[test]
+    zope.testrunner
```
Isn't zope.testrunner already part of the .[test] extra?  You don't install it explicitly in `.travis.yml`.

mgedmin reviewed (approved):
LGTM with minor possible improvements.

mgedmin commented:
(One other thing I would do before the next release is rename *.txt to *.rst for prettier GitHub rendering.)

janwijbrand committed:
cleanups

janwijbrand committed:
prettier text file rendering
